syncano/models/custom_sockets_utils.py

```diff
@@ -0,0 +1,250 @@
+    """
+    The type of the dependency object used in the custom socket;
+    """
+    SCRIPT = 'script'
+
+
+class BaseCall(object):
+    """
+    Base class for call object.
+    """
+
+    call_type = None
+
+    def __init__(self, name, methods):
+        self.name = name
+        self.methods = methods
+
+    def to_dict(self):
+        if self.call_type is None:
+            raise SyncanoValueError('call_type not set.')
+        return {
+            'type': self.call_type,
+            'name': self.name,
+            'methods': self.methods
+        }
+
+
+class ScriptCall(BaseCall):
+    """
+    Script call object.
+
+    The JSON format is as follows (to_dict in the base class)::
+
+        {
+            'type': 'script',
+            'name': '<script_label>,
+            'methods': [<method_list>],
+        }
+
+    methods can be as follows:
+        * ['GET']
+        * ['*'] - which will do a call on every request method;
+    """
+    call_type = CallType.SCRIPT
+
+
+class Endpoint(object):
+    """
+    The object which stores metadata about endpoints in custom socket;
+
+    The JSON format is as follows::
+
+        {
+            '<endpoint_name>: {
+                'calls': [
+                    <list of JSON format of Calls objects>
+                ]
+            }
+        }
+
+    """
+    def __init__(self, name):
+        self.name = name
+        self.calls = []
+
+    def add_call(self, call):
+        self.calls.append(call)
+
+    def to_endpoint_data(self):
+        return {
+            self.name: {
+                'calls': [call.to_dict() for call in self.calls]
+            }
+        }
+
+
+class BaseDependency(object):
+    """
+    Base dependency object;
+
+    On the base of the fields attribute - the JSON format of the dependency is returned.
+    The fields are taken from the dependency object - which can be Script (supported now).
+    """
+
+    fields = []
+    dependency_type = None
+
+    def to_dependency_data(self):
+        if self.dependency_type is None:
+            raise SyncanoValueError('dependency_type not set.')
+        dependency_data = {'type': self.dependency_type}
+        dependency_data.update(self.get_dependency_data())
+        return dependency_data
+
+    def get_name(self):
+        raise NotImplementedError()
+
+    def get_dependency_data(self):
+        raise NotImplementedError()
+
+    def create_from_raw_data(self, raw_data):
+        raise NotImplementedError()
+
+
+class ScriptDependency(BaseDependency):
+    """
+    Script dependency object;
+
+    The JSON format is as follows::
+        {
+            'type': 'script',
+            'runtime_name': '<runtime name defined in RuntimeChoices>',
+            'source': '<source>',
+            'name': '<name>'
+        }
+    """
+
+    dependency_type = DependencyType.SCRIPT
+    fields = [
+        'runtime_name',
+        'source'
+    ]
+
+    def __init__(self, name=None, script=None, script_endpoint=None):
```

Comment:
It doesn't make sense to have both script and script_endpoint defined here.
So something like:
`def __init__(self, script_or_script_endpoint, name=None):`
is a little easier to use.
Then:
```
ScriptDependency(
    Script.please.first(),
    name='custom_script'
)  # OK

ScriptDependency(
    ScriptEndpoints.please.first()
) # OK, use name from ScriptEndpoint

ScriptDependency(
    ScriptEndpoints.please.first(),
    name='custom_script'
) # OK, override script endpoint's name

ScriptDependency(
    Script.please.first()
) # NOT OK. name is required if first param is of instance Script
```